larray/core/array.py

```diff
@@ -5572,6 +5542,201 @@ def extend(self, axis, other):
+        >>> arr1.insert(42, before='b1', label='b0.5')
+        a\\b  b0  b0.5  b1  b2
+         a0   0    42   1   2
+         a1   3    42   4   5
+        >>> arr2 = ndtest(2)
+        >>> arr2
+        a  a0  a1
+            0   1
+        >>> arr1.insert(arr2, after='b0', label='b0.5')
+        a\\b  b0  b0.5  b1  b2
+         a0   0     0   1   2
+         a1   3     1   4   5
+        >>> arr1.insert(42, axis='b', pos=1, label='b0.5')
+        a\\b  b0  b0.5  b1  b2
+         a0   0    42   1   2
+         a1   3    42   4   5
+        >>> arr1.insert(42, before=X.b.i[1], label='b0.5')
+        a\\b  b0  b0.5  b1  b2
+         a0   0    42   1   2
+         a1   3    42   4   5
+
+        insert an array which already has the axis
+
+        >>> arr3 = ndrange('a=a0,a1;b=b0.1,b0.2') + 42
+        >>> arr3
+        a\\b  b0.1  b0.2
+         a0    42    43
+         a1    44    45
+        >>> arr1.insert(arr3, before='b1')
+        a\\b  b0  b0.1  b0.2  b1  b2
+         a0   0    42    43   1   2
+         a1   3    44    45   4   5
+        """
+
+        # XXX: unsure we should have arr1.insert(arr3, before='b1,b2') result in (see unit tests):
+
+        # a\\b  b0  b0.1  b1  b0.2  b2
+        #  a0   0    42   1    43   2
+        #  a1   3    44   4    45   5
+
+        # we might to implement the following instead:
+
+        # a\\b  b0  b0.1  b0.2  b1  b0.1  b0.2  b2
+        #  a0   0    42    43   1    42    43   2
+        #  a1   3    44    45   4    44    45   5
+
+        # The later looks less useful and could be emulated easily via:
+        # arr1.insert([arr3, arr3], before='b1,b2')
+        # while the above is a bit harder to achieve manually:
+        # arr1.insert([arr3[[b]] for b in arr3.b], before=['b1', 'b2'])
+        # but the later is *probably* more intuitive (and wouldn't suffer from the inefficiency we currently have).
+
+        # XXX: when we have several lists, we implicitly match them by position, which we should avoid for the usual
+        # reason, but I am unsure what the best syntax for that would be.
+
+        # the goal is to get this result
+
+        # a\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # When the inserted arrays already contain a label, this seems reasonably readable:
+
+        # >>> arr1 = ndtest((2, 3))
+        # >>> arr1
+        # a\\b  b0  b1  b2
+        #  a0   0   1   2
+        #  a1   3   4   5
+        # >>> arr2 = full('b=b0.5', 8)
+        # >>> arr2
+        # b  b0.5
+        #       8
+        # >>> arr3 = full('b=b1.5', 9)
+        # >>> arr3
+        # b  b1.5
+        #       9
+        # >>> arr1.insert(before={'b1': arr2, 'b2': arr3})
+        # a\\b  b0  b0.5  b1  b1.5  b2
+        #  a0   0     8   1     9   2
+        #  a1   3     8   4     9   5
+
+        # When the inserted arrays/values have no label, this does not really convince me and it prevents using after
```

Comment:
or we could ask to pass 3 lists: 1 for value, 1 for before or after and 1 for label (they must have the same length).

Reply:
that is precisely what is already implemented and what I think we should avoid, because it matches items by position within each list, which for longer lists is dangerous (for the same reason stack(list_of_values, axis) is not recommended): it will let silly inversion errors pass silently and those will be very hard to find.